Add mute/unmute (penalty) functionality

davidlougheed opened the pull request:
Adds a mute function with the ability to cache/restore roles and apply a penalty role

## Description
`?mute` places a user in penalty and removes all their other roles, storing them in a db table
`?unmute` removes user from penalty and restores their roles

## Motivation and Context
penalty is annoying to do by hand, and role overrides can lead to issues - this automates away that problem

## How Has This Been Tested?
tested on idoneam server (still more to be done, WIP)

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

davidlougheed committed:
Add penalty role config option

davidlougheed committed:
Add ?mute and ?unmute for sending users to the penalty box
Also move previous roles command to roles.py

davidlougheed committed:
Create separate table for penalty roles
Refactor out shared code from ?previous_roles

davidlougheed committed:
Yapf

davidlougheed committed:
Misc other roles tweaks
(unrelated stuff, sorry catan/potate)

davidlougheed committed:
y a p f
a p f y
p f y a
f y a p

davidlougheed committed:
Add penalty role config to readme

davidlougheed committed:
Merge remote-tracking branch 'origin/dev' into mute-function

# Conflicts:
#	cogs/roles.py

davidlougheed committed:
Make in_channel command name consistent w/ others

le-potate commented on Martlet.schema:
```diff
@@ -49,7 +49,15 @@ CREATE TABLE IF NOT EXISTS `BankTransactions` (
+-- Used for when users are placed in the Penalty role; stores old roles akin to
+-- the way PreviousRoles does it.
+CREATE TABLE IF NOT EXISTS `PenaltyUsers` (
+    `ID`	      INTEGER UNIQUE,
+	`Roles`       TEXT -- space-separated list of role IDs
```
seems like theres a mix of tabs and spaces there, but it's not the only place like that in the file

davidlougheed replied:
560f8ee

le-potate commented on README.md:
```diff
@@ -52,6 +52,8 @@ You must set certain values in the `config.ini` file, in particular your Discor
     * `DownvoteEmoji`: The name of your downvote emoji.
     * `ModeratorRole`: The name of the role that your moderators have (for functions like DMing users).
     * `DeveloperRole`: The name of the role that your developers have (for functions like restarting the bot). This could be the same role than moderator.
+    * `PenaltyRole`: The name of the role used for placing users in a "penalty box" (which presumably removes some 
```
i think this should be optional. right now, if you dont have the penaltyrole, you get the exception mis-configured penalty role everytime someone leaves

davidlougheed replied:
06e0fe4

le-potate commented on cogs/roles.py:
```diff
@@ -257,8 +264,238 @@ async def create_role(self, ctx, *, role: Optional[str] = None):
+            c = conn.cursor()
+            penalty = c.execute("SELECT * FROM PenaltyUsers WHERE ID = ?",
+                                (user.id, )).fetchone()
+            return penalty is None
```
doesnt this do the opposite of what it's supposed to. If penality is none, then is_in_penalty_table should return False, not True. Right now, every new user gets the penalty role !

davidlougheed replied:
🐒

davidlougheed replied:
678a97a

le-potate commented on cogs/roles.py:
```diff
@@ -257,8 +264,238 @@ async def create_role(self, ctx, *, role: Optional[str] = None):
+            c = conn.cursor()
+            # store roles as a string of IDs separated by spaces
+            t = (user.id, " ".join(str(e) for e in roles_id))
+            c.execute(f"REPLACE INTO {table} VALUES (?, ?)", t)
```
If a user gets muted but has no roles, they wont get added to the penalty table

davidlougheed replied:
9dd2412

davidlougheed committed:
Use spaces consistently in schema file

davidlougheed committed:
Fix penalized users not getting an entry if they don't have roles

davidlougheed committed:
Roles: make penalty stuff optional

davidlougheed committed:
Roles: Fix inverted logic for penalty stuff

davidlougheed committed:
Schema: fix some alignment stuff

davidlougheed committed:
Merge branch 'dev' into mute-function

davidlougheed committed:
yet another popflash

davidlougheed committed:
Merge remote-tracking branch 'origin/dev' into mute-function

# Conflicts:
#	Martlet.schema
#	cogs/helpers.py
#	cogs/roles.py

davidlougheed committed:
Add loading message for restoring roles

davidlougheed committed:
Bump some copyrights

davidlougheed committed:
Merge dev

davidlougheed committed:
merge fixes

le-potate committed:
Merge branch 'dev' into mute-function

le-potate committed:
formatting

le-potate committed:
more formatting

sorry i dont know why my linter isnt working

le-potate committed:
more formatting 2

lmao

le-potate committed:
Fix typo

le-potate committed:
fix infinite loop and inverted logic

also made sure that the bot will add the penalty role even if it cannot remove all roles

le-potate committed:
formatting

le-potate reviewed (approved):
I am currently working on a new muting/appeal system that will work a bit differently, but still use a lot of this. This is working so it's fine if the bot ends up being updated before I'm finished with the new system, but I think I will be done this week.

davidlougheed commented:
i am a god of code
